File: loopx/control_plane/coordination/coordination_state_contract.ts
import { readFileSync } from "node:fs";

import type { JsonObject } from "../effect_program.ts";
import {
  AuthorityStoreProtocolError,
  canonicalAuthorityObject,
} from "./authority_store_codec.ts";

export const COORDINATION_STATE_CONTRACT_SCHEMA =
  "loopx_coordination_state_contract_v0";

interface RecordContract {
  readonly fields: readonly string[];
  readonly required_fields: readonly string[];
}

interface TodoRecordContract extends RecordContract {
  readonly schema_version: string;
  readonly item_schema_version: string;
}

interface CoordinationStateContract {
  readonly schema_version: typeof COORDINATION_STATE_CONTRACT_SCHEMA;
  readonly todo_read_record: TodoRecordContract;
  readonly compatibility: {
    readonly unknown_field_policy: "reject";
    readonly field_removal_policy: "maintainer_approval_required";
    readonly markdown_role: "human_workbench_and_compatibility_projection";
  };
}

function stringList(value: unknown, label: string): readonly string[] {
  if (!Array.isArray(value) || value.length === 0 ||
      value.some((item) => typeof item !== "string" || item.length === 0)) {
    throw new AuthorityStoreProtocolError(`${label} must contain non-empty strings`);
  }
  if (new Set(value).size !== value.length) {
    throw new AuthorityStoreProtocolError(`${label} must not contain duplicates`);
  }
  return Object.freeze([...value] as string[]);
}

function recordContract(value: unknown, label: string): RecordContract {
  const record = canonicalAuthorityObject(value, label);
  return Object.freeze({
    fields: stringList(record.fields, `${label}.fields`),
    required_fields: stringList(record.required_fields, `${label}.required_fields`),
  });
}

function loadCoordinationStateContract(): CoordinationStateContract {
  const raw = canonicalAuthorityObject(JSON.parse(readFileSync(
    new URL("./coordination_state_contract_v0.json", import.meta.url),
    "utf8",
  )), "coordination state contract");
  if (raw.schema_version !== COORDINATION_STATE_CONTRACT_SCHEMA) {
    throw new AuthorityStoreProtocolError("coordination state contract schema mismatch");
  }
  const todo = canonicalAuthorityObject(raw.todo_read_record, "todo_read_record");
  const compatibility = canonicalAuthorityObject(raw.compatibility, "compatibility");
  if (typeof todo.schema_version !== "string" ||
      typeof todo.item_schema_version !== "string" ||
      compatibility.unknown_field_policy !== "reject" ||
      compatibility.field_removal_policy !== "maintainer_approval_required" ||
      compatibility.markdown_role !== "human_workbench_and_compatibility_projection") {
    throw new AuthorityStoreProtocolError("coordination state contract policy mismatch");
  }
  return Object.freeze({
    schema_version: COORDINATION_STATE_CONTRACT_SCHEMA,
    todo_read_record: Object.freeze({
      ...recordContract(todo, "todo_read_record"),
      schema_version: todo.schema_version,
      item_schema_version: todo.item_schema_version,
    }),
    compatibility: Object.freeze({
      unknown_field_policy: "reject",
      field_removal_policy: "maintainer_approval_required",
      markdown_role: "human_workbench_and_compatibility_projection",
    }),
  });
}

export const COORDINATION_STATE_CONTRACT = loadCoordinationStateContract();
export const TODO_CANONICAL_READ_RECORD_SCHEMA =
  COORDINATION_STATE_CONTRACT.todo_read_record.schema_version;
export const TODO_ITEM_SCHEMA =
  COORDINATION_STATE_CONTRACT.todo_read_record.item_schema_version;
export const TODO_CANONICAL_READ_RECORD_FIELDS =
  COORDINATION_STATE_CONTRACT.todo_read_record.fields;
export const TODO_CANONICAL_REQUIRED_READ_FIELDS =
  COORDINATION_STATE_CONTRACT.todo_read_record.required_fields;

export function canonicalCoordinationRecord(
  value: unknown,
  contract: RecordContract,
  label: string,
): JsonObject {
  const record = canonicalAuthorityObject(value, label);
  const allowed = new Set(contract.fields);
  const unexpected = Object.keys(record)
    .filter((field) => !allowed.has(field))
    .sort((left, right) => left.localeCompare(right));
  if (unexpected.length > 0) {
    throw new AuthorityStoreProtocolError(
      `${label} has unversioned fields: ${unexpected.join(", ")}`,
    );
  }
  const missing = contract.required_fields.filter((field) => !(field in record));
  if (missing.length > 0) {
    throw new AuthorityStoreProtocolError(
      `${label} omits required fields: ${missing.join(", ")}`,
    );
  }
  return canonicalAuthorityObject(Object.fromEntries(
    contract.fields.flatMap((field) => field in record ? [[field, record[field]]] : []),
  ), label);
}

export function canonicalCoordinationTodoRecord(
  value: unknown,
  label = "coordination Todo read record",
): JsonObject {
  const record = canonicalCoordinationRecord(
    value,
    COORDINATION_STATE_CONTRACT.todo_read_record,
    label,
  );
  if (record.schema_version !== TODO_ITEM_SCHEMA ||
      (record.role !== "user" && record.role !== "agent") ||
      typeof record.status !== "string" || record.status.length === 0 ||
      typeof record.done !== "boolean" || typeof record.text !== "string" ||
      typeof record.archive_state !== "string" || record.archive_state.length === 0 ||
      typeof record.source_section !== "string" || record.source_section.length === 0) {
    throw new AuthorityStoreProtocolError(`${label} has invalid required semantics`);
  }
  return record;
}

File: loopx/control_plane/coordination/coordination_state_contract.py
"""Versioned coordination-state record contract shared with TypeScript.

The JSON artifact is the language-neutral schema owner.  Python projection
code may normalize Markdown into this shape, but it must not silently invent a
different field allowlist or discard an unversioned machine-owned field.
"""

from __future__ import annotations

import json
from importlib.resources import files
from typing import Any, Mapping


COORDINATION_STATE_CONTRACT_SCHEMA_VERSION = "loopx_coordination_state_contract_v0"
_CONTRACT_RESOURCE = "coordination_state_contract_v0.json"


class CoordinationStateContractError(ValueError):
    """A record or bundled contract violates the coordination-state schema."""


def _string_tuple(value: object, *, label: str) -> tuple[str, ...]:
    if not isinstance(value, list) or not value:
        raise CoordinationStateContractError(f"{label} must be a non-empty array")
    if any(not isinstance(item, str) or not item for item in value):
        raise CoordinationStateContractError(f"{label} must contain non-empty strings")
    if len(set(value)) != len(value):
        raise CoordinationStateContractError(f"{label} must not contain duplicates")
    return tuple(value)


def _validate_compatibility(value: object) -> None:
    if not isinstance(value, dict) or value != {
        "unknown_field_policy": "reject",
        "field_removal_policy": "maintainer_approval_required",
        "markdown_role": "human_workbench_and_compatibility_projection",
    }:
        raise CoordinationStateContractError("coordination state contract policy mismatch")


def _load_contract() -> dict[str, Any]:
    raw = json.loads(
        files("loopx.control_plane.coordination")
        .joinpath(_CONTRACT_RESOURCE)
        .read_text(encoding="utf-8")
    )
    if not isinstance(raw, dict) or raw.get("schema_version") != (
        COORDINATION_STATE_CONTRACT_SCHEMA_VERSION
    ):
        raise CoordinationStateContractError("coordination state contract schema mismatch")
    _validate_compatibility(raw.get("compatibility"))
    return raw


COORDINATION_STATE_CONTRACT = _load_contract()
_TODO_READ_RECORD = COORDINATION_STATE_CONTRACT.get("todo_read_record")
if not isinstance(_TODO_READ_RECORD, dict):
    raise CoordinationStateContractError("Todo record contract must be an object")

TODO_CANONICAL_READ_RECORD_SCHEMA_VERSION = _TODO_READ_RECORD.get("schema_version")
TODO_ITEM_SCHEMA_VERSION = _TODO_READ_RECORD.get("item_schema_version")
if (
    not isinstance(TODO_CANONICAL_READ_RECORD_SCHEMA_VERSION, str)
    or not TODO_CANONICAL_READ_RECORD_SCHEMA_VERSION
    or not isinstance(TODO_ITEM_SCHEMA_VERSION, str)
    or not TODO_ITEM_SCHEMA_VERSION
):
    raise CoordinationStateContractError("Todo record schema versions must be strings")
TODO_CANONICAL_READ_RECORD_FIELDS = _string_tuple(
    _TODO_READ_RECORD.get("fields"), label="todo_read_record.fields"
)
TODO_CANONICAL_REQUIRED_READ_FIELDS = _string_tuple(
    _TODO_READ_RECORD.get("required_fields"),
    label="todo_read_record.required_fields",
)
_UNKNOWN_REQUIRED_FIELDS = sorted(
    set(TODO_CANONICAL_REQUIRED_READ_FIELDS).difference(TODO_CANONICAL_READ_RECORD_FIELDS)
)
if _UNKNOWN_REQUIRED_FIELDS:
    raise CoordinationStateContractError(
        "todo_read_record.required_fields are absent from fields: "
        + ", ".join(_UNKNOWN_REQUIRED_FIELDS)
    )

_DOMAIN = COORDINATION_STATE_CONTRACT["todo_domain_record"]
_PROJECTION = COORDINATION_STATE_CONTRACT["todo_projection_metadata"]
if (
    _DOMAIN.get("fields_from") != "todo_read_record"
    or _DOMAIN.get("exclude_fields_from") != "todo_projection_metadata"
):
    raise CoordinationStateContractError("Todo domain contract mismatch")
TODO_PROJECTION_METADATA_FIELDS = _string_tuple(
    _PROJECTION.get("fields"), label="todo_projection_metadata.fields"
)
TODO_DOMAIN_READ_RECORD_SCHEMA_VERSION: str = _DOMAIN["schema_version"]
TODO_DOMAIN_ITEM_SCHEMA_VERSION: str = _DOMAIN["item_schema_version"]
TODO_DOMAIN_RECORD_FIELDS = tuple(
    field for field in TODO_CANONICAL_READ_RECORD_FIELDS
    if field not in TODO_PROJECTION_METADATA_FIELDS
)
TODO_DOMAIN_REQUIRED_FIELDS = _string_tuple(
    _DOMAIN.get("required_fields"), label="todo_domain_record.required_fields"
)
if not set(TODO_DOMAIN_REQUIRED_FIELDS) <= set(TODO_DOMAIN_RECORD_FIELDS):
    raise CoordinationStateContractError("Todo domain required fields are not declared")


def canonical_record_fields(
    value: Mapping[str, object],
    *,
    fields: tuple[str, ...],
    required_fields: tuple[str, ...],
    label: str,
    reject_unknown: bool,
) -> dict[str, object]:
    """Copy one record through the shared field contract.

    ``reject_unknown`` is mandatory for provider-bound projections.  The false
    mode exists only for legacy status callers that may carry presentation-only
    fields outside the machine-owned coordination record.
    """

    unknown_required = sorted(set(required_fields).difference(fields))
    if unknown_required:
        raise CoordinationStateContractError(
            f"{label} required fields are absent from fields: "
            + ", ".join(unknown_required)
        )
    if reject_unknown:
        unexpected = sorted(set(value).difference(fields))
        if unexpected:
            raise CoordinationStateContractError(
                f"{label} has unversioned fields: {', '.join(unexpected)}"
            )
    record = {
        field: value[field]
        for field in fields
        if field in value and value[field] is not None
    }
    missing = [field for field in required_fields if field not in record]
    if missing:
        raise CoordinationStateContractError(
            f"{label} omits required fields: {', '.join(missing)}"
        )
    return record


__all__ = [
    "COORDINATION_STATE_CONTRACT",
    "COORDINATION_STATE_CONTRACT_SCHEMA_VERSION",
    "CoordinationStateContractError",
    "TODO_CANONICAL_READ_RECORD_FIELDS",
    "TODO_CANONICAL_READ_RECORD_SCHEMA_VERSION",
    "TODO_CANONICAL_REQUIRED_READ_FIELDS",
    "TODO_ITEM_SCHEMA_VERSION",
    "TODO_DOMAIN_READ_RECORD_SCHEMA_VERSION",
    "TODO_DOMAIN_ITEM_SCHEMA_VERSION",
    "TODO_DOMAIN_RECORD_FIELDS",
    "TODO_DOMAIN_REQUIRED_FIELDS",
    "TODO_PROJECTION_METADATA_FIELDS",
    "canonical_record_fields",
]
